packages/sdk/src/oracle/services/statemachines/approve.ts

```diff
@@ -0,0 +1,56 @@
+import assert from "assert";
+import { Update } from "../update";
+import Store from "../../store";
+import { Signer } from "../../types/ethers";
+import { Handlers as GenericHandlers } from "../../types/statemachine";
+import { ContextClient } from "./utils";
+
+export type Params = {
+  currency: string;
+  chainId: number;
+  signer: Signer;
+  account: string;
+  spender: string;
+  amount: string;
+  confirmations: number;
+};
+
+export type Memory = { hash?: string };
+
+export function initMemory(): Memory {
+  return {};
+}
+
+export function Handlers(store: Store): GenericHandlers<Params, Memory> {
+  const update = new Update(store);
+  return {
+    async start(params: Params, memory: Memory) {
+      const { chainId, currency, spender, amount, account, signer } = params;
+      assert(chainId === (await signer.getChainId()), "Signer on wrong chainid");
+      assert(account === (await signer.getAddress()), "Signer on wrong account");
+
+      // create service if it does not exist
+      store.write((w) => w.services(chainId).erc20s(currency));
+      const erc20 = store.read().tokenService(chainId, currency);
+      const tx = await erc20.approve(signer, spender, amount);
```

Comment:
theres a slight wrinkle here in that waiting on the users interaction here blocks the statemachine queue. This is neither good nor bad, but it implies that we cannot change state ( ie issue new transactions or change user) until the user closes or accepts this popup in metamask. I suggest keeping this behavior unless we see this is detrimental to the application. We could easily introduce a timeout here in case the user forgets to proceed.

Reply:
currently metamask will force you to resolve the transaction request before you can change user, so it doesnt matter too much. there are some cases like "disconnect" which you could do while waiting for a tx, in this case the disconnect wont happen until you acknowledge a tx request in MM.  I think its best to show the user an error if this happens saying why they cant disconnect, ie they must accept or cancel the pending transaction.